### DISCUSSION TITLE
Custom Fabric.js line with controls

### DISCUSSION THREAD

**Discussion by taherPardawala** [fabricjs/fabric.js#7940] [Show and tell]
Hi and thanks for developing such a great drawing library!

I have been working on Fabric with my colleague Huzefa for the last 6 months as we are working on a project similar to the Shutterstock Editor. During this time we faced many challenges one of them included creating Lines with custom controls.

Since we figured out a solution for our problem we decided to pen it down and publish it for other aspiring Fabric devs who might want to implement something similar.

Here is the link to the article - [LINK](https://order-stack.medium.com/a-better-line-custom-fabric-js-line-with-controls-f5e0e311579e). Would be glad if you could give it a read.

Looking forward to any feedback. Have a great day!

**Comment by ShaMan123** [comment]
Looks good.
Line is supposed to be a simple object so you would expect it to be easy to customize but in truth it's a hazard to do so.
That's why I hate it and stay away from it and use Polygon/Polyline.
We are considering the deprecation of Line, Polyline et al. in favor of Path and solid utils.
I've played around with the codepen, feels good.
I suggest leveraging the Control API which is designed exactly for this purpose. It might save up some event handlers and redundancies as seen here http://fabricjs.com/custom-controls-polygon

**Reply by taherPardawala** [reply]
Thanks for the feedback @ShaMan123 

You are right we could have probably leveraged the controls API to avoid a few of the things we are doing here. When we first worked on this implementation we were not aware of it and hence we went with this approach.

But yes it is worth looking into the controls API for this implementation. Maybe we can do a V2 of this in some time.

**Reply by graphics-et-al** [reply]
Excellent- I'm currently struggling with curves. An implementation of what I have so far can be found at https://forum.dev.graphicsetal.com/fabricjs/curve.html

It works fine, but what I'm struggling with is when they're drawn by clicking and dragging- setting the points doesn't update the width and height for some reason.

**Reply by taherPardawala** [reply]
Hey @graphics-et-al , can you give me some more details about the issue. I cannot grasp the problem from the demo you shared.
